### PR TITLE
`LocalCall`

### PR DESCRIPTION
**Pull request by joan41868**
**Description**:

This **PR** provides implementation of the `signer.call` method in the `Wallet` class. 
Added:
- `getHederaClient()` in `provider`
- `call()` in `Wallet`

Deprecated call in abstract signer - now throws

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by joan41868** [comment]
Let's merge `contract-interactions` in develop before reviewing this, as `local-call` is based on it.

File: packages/wallet/src.ts/index.ts [hunk 1]
```diff
@@ -281,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
 		return encryptKeystore(this, password, options, progressCallback);
 	}
 
+	async call(txRequest : Deferrable<TransactionRequest>, blockTag?: BlockTag | Promise<BlockTag>): Promise<string> {
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:292]: We s hould remove the `blockTag`

File: packages/wallet/src.ts/index.ts [hunk 2]
```diff
@@ -220,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
...
+		this._checkProvider("call");
+		const tx = await resolveProperties(this.checkTransaction(txRequest));
+		const contractAccountLikeID = getAccountFromAddress(tx.to.toString());
+		const contractId = `${contractAccountLikeID.shard}.${contractAccountLikeID.realm}.${contractAccountLikeID.num}`;
+
+		const thisAcc = getAccountFromAddress(await this.getAddress());
+		const thisAccId = `${thisAcc.shard}.${thisAcc.realm}.${thisAcc.num}`;
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:299]: Lets implement new method in `address/index.ts` : `asAccountString(AccountLike)` which will expose the logic

File: packages/wallet/src.ts/index.ts [hunk 3]
```diff
@@ -220,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
...
+
+		const nodeID = AccountId.fromString(tx.nodeId.toString());
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:301]: nodeId is `AccountLike` thus it must be treated as `contractId` and `thisAccount` -> using the newly created util function for parsing AccontLike to account string

File: packages/wallet/src.ts/index.ts [hunk 4]
```diff
@@ -220,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
...
+		const paymentTxId = TransactionId.generate(thisAccId);
+
+		const hederaTx = new ContractCallQuery()
+			.setContractId(contractId)
+			.setFunctionParameters(arrayify(tx.data))
+			.setNodeAccountIds([nodeID])
+			.setGas(Long.fromString(tx.gasLimit.toString()))
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:308]: Lets use the `numberify` method instead and remove the `long` dependency

File: packages/wallet/src.ts/index.ts [hunk 5]
```diff
@@ -220,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
...
+				transfers: {
+					accountAmounts:[
+						{
+							accountID: AccountId.fromString(thisAccId)._toProtobuf(),
+							amount: new Hbar(3).negated().toTinybars()
+						},
+						{
+							accountID: nodeID._toProtobuf(),
+							amount: new Hbar(3).toTinybars()
+						}
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:328]: Lets add `TODO` here that the exact amount will be computed from `getCost` once `getCost` works without operator

File: packages/wallet/src.ts/index.ts [hunk 6]
```diff
@@ -220,6 +289,68 @@ export class Wallet extends Signer implements ExternallyOwnedAccount, TypedDataS
...
+			sigPair: [walletKey.publicKey._toProtobufSignature(signature)]
+		}
+
+		const transferSignedTransactionBytes =  SignedTransaction.encode(signed).finish();
+		hederaTx._paymentTransactions.push({
+			signedTransactionBytes: transferSignedTransactionBytes
+		});
+		const response = await hederaTx.execute(this.provider.getHederaClient());
+		// TODO: this may not be the best thing to return but it should work for testing
+		return hexlify(response.asBytes());
```
Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:351]: We should return the `contractCallResult` bytes and not the whole `ContractFunctionResult` object:
image

Meaning that we only need to return the `return` from the solidity function as it is performed in `ethers`

Comment (Daniel-K-Ivanov) [inline packages/wallet/src.ts/index.ts:351]: We must parse situations in which the contract call reverts for some reason. See the ethers implementation for reference. In the `provider.call` method if the response code is > 400 exception is thrown and the exception is handled through this:
https://github.com/ethers-io/ethers.js/blob/0578a88efa4f5f65df90a5ac8b63ffdfb5e44d9a/packages/providers/lib/json-rpc-provider.js#L68

We need to have similar experience to failing queries or reverting TX

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]

File: packages/wallet/src.ts/index.ts (hunk 6 above)
Comment (joan41868) [reply packages/wallet/src.ts/index.ts]: I think there's an issue with this logic - the only `abstract` property of the call response are the bytes. 
We've got the following options as well:
options1
options2

Which one would you chose?

File: packages/address/src.ts/index.ts [hunk 7]
```diff
@@ -18,6 +18,14 @@ import {version} from "./_version";
 
 const logger = new Logger(version);
 
+export function asAccountString(acc :AccountLike): string {
+    if (typeof acc === "string") {
+        return acc;
+    } else {
+        return `${acc.shard}${acc.realm}${acc.num}`;
```
Comment (joan41868) [inline packages/address/src.ts/index.ts:25]: add dots

**Pull request merged**